lib/sprite_batch.dart

```diff
@@ -2,66 +2,244 @@ import 'dart:ui';
+      images: images,
+    );
+  }
+}
+
+/// A single item in a SpriteBatch.
+///
+/// Holds all the important information of a batch item,
+///
+/// Web currently does not support `Canvas.drawAtlas`, so a BatchItem will
+/// automatically calculate a transform matrix based on the [transform] value, to be
+/// used when rendering on the web. It will initialize a [destination] object
+/// and a [paint] object.
+class BatchItem {
+  /// The source rectangle on the [SpriteBatch.atlas].
+  final Rect source;
+
+  /// The destination rectangle for the Canvas.
+  ///
+  /// It will be transformed by [matrix].
+  final Rect destination;
+
+  /// The transform values for this batch item.
+  final RSTransform transform;
+
+  /// The background color for this batch item.
+  final Color color;
+
+  /// Fallback matrix for the web.
+  ///
+  /// Because `Canvas.drawAtlas` is not supported on the web we also
+  /// build a `Matrix4` based on the [transform] values.
+  final Matrix4 matrix;
+
+  /// Paint object used for the web.
+  final Paint paint;
+
+  BatchItem({
+    @required this.source,
+    @required this.transform,
+    @required this.color,
+  })  : assert(source != null),
+        assert(transform != null),
+        assert(color != null),
+        matrix = Matrix4(
+          transform.scos,
+          transform.ssin,
+          0,
+          0,
+          -transform.ssin,
+          transform.scos,
+          0,
+          0,
+          0,
+          0,
+          0, // <-- this is the scale value, we can't determine this from a RSTransform,
+          // but we also don't need to do s because it is already calculated
+          // inside the transform values.
+          0,
+          transform.tx,
+          transform.ty,
+          0,
+          1,
+        ),
+        paint = Paint()..color = color,
+        destination = Offset.zero & source.size;
+}
+
+/// The SpriteBatch API allows for rendering multiple items at once.
+///
+/// This class allows for optimization when you want to draw many parts of an
+/// image onto the canvas. It is more efficient than using multiple calls to [drawImageRect]
+/// and provides more functionality by allowing each [BatchItem] to have their own transform
+/// rotation and color.
+///
+/// By collecting all the necessary transforms on a single image and sending those transforms
+/// in a single batch to the GPU, we can render multiple parts of a single image at once.
+///
+/// **Note**: Currently web does not support `Canvas.drawAtlas`, which SpriteBatch uses under
+/// the hood, instead it will render each [BatchItem] using `Canvas.drawImageRect`, so there
+/// might be a performance hit on web when working with many batch items.
 class SpriteBatch {
+  /// List of all the existing batch items.
+  final _batchItems = <BatchItem>[];
+
+  /// The sources to use on the [atlas].
+  final _sources = <Rect>[];
+
+  /// The transforms that should be applied on the [_sources].
+  final _transforms = <RSTransform>[];
```

Comment:
After working a bit more with the API, I came to the conclusion that exposing both the `_transforms` and `_sources` list to the public can be quite useful for optimizing rendering.

My current usecase is that I reuse the transforms modify them slightly and then pass them back to the batch through `addTransform`. That way it won't have to do any kind of calculation, as my modification is just a simple `+` and `-`. 

And if we don't expose these lists then a user will have to precreate his own `RSTransform`s instead of using our helper method `add`.

Reply:
But is it safe to directly change the transformations?

Reply:
It is as safe as passing the wrong values to the addTransform method ;-). So I would not recommend it, but it does allow people to use the SpriteBatch more efficiently when they know what they are doing

Otherwise you'll always have the calculation overhead. And you do notice it with many transforms

Reply:
Sounds good to me then!

Reply:
Then I'll refactor this PR a little bit more soon